Implemented Full Coverage Testing for ConfidenceScoreOptimizer, ConfidenceScoreCalculator, ButtonStep, CardStep

auroraborie opened the pull request:
amended ConfidenceScoreOptimizer to return the ranked talksteps by their confidence score in decreasing order as an ArrayList of Maps. Made the corresponding changes to WozWozTranscriptAnalysisInteractor to be compatible with these changes so that the same functionality remains. 

-A capping feature was implemented on ConfidenceScoreCalculator so no percentages exceeding greater than 100 was returned
-ConfidenceScoreCalculator also now rounds the confidence score to 2 decimal places.

Notes for the Testing:
-full coverage testing for every line for the entities ButtonStep, CardStep, ConfidenceScoreOptimizer, ConfidenceScoreCalculator
This was done with a variety of dialogues suitable for specific TalkSteps to demonstrate suitability. 
-max scores for ButtonStep and CardStep were amended to provide more accurate results

auroraborie committed:
implemented TestSuite for ButtonStep

auroraborie committed:
Merge remote tracking branch 'origin' into managertests2

auroraborie committed:
amended ConfidenceScoreOptimizer to return the ranked talksteps by their confidence score in decreasing order as an ArrayList of Maps.

auroraborie committed:
fixe some Maven bugs

auroraborie committed:
fixe some Maven bugs

Lemiex reviewed (changes requested):
Thanks for being so speedy on changing the Optimizer!!
1. you mightve accidentally changed TalkStep, but don't alter it! The correct version is already merged with main, see if you can restore the 'break' that you deleted and push again. If that makes the TalkStep change go away in this PR, great, if not, go to the TalkStep currently on our github and just copy paste the whole file and push it again so it's the right version
2. Update the documentation in the Interactor line 77, since we're returning all the talksteps now
3. Super minimal and up to you but i thought for Optimzier, rankSuggestions() might be a better method name than findSuggestedTalkSteps(), since we aren't rly "finding" them
4. I like your BeforeEach init(). method, it's concise. i was trying to write somthing similar and it was like so long TAT
5. ButtonSteptests - i appreicate your work for testUserMsgLength() on line 94 but it's not a responsibility for ButtonStep. for clarity of class responsibility, gotta leave all those tests to TalkStep
6. ButtonSteptests - what is line 34-46 the keyword lists for? you don't use them in the tests
7. ButtonStepTests - you might be missing tests for getMaxScore, getScoreAccumulator, and runAnalysis()
8. For methods like runAnalysis(), getScoreAccumulator(), setZeroAccumulator(), you might wanna play with some combination since they depend on runAnalysis() already happening
9. When you test runAnalysis(), use mock functions for countMatchKeywords() and calculateMsgLength(). Specific implementation: ` TalkStep talkStep = mock(TalkStep.class);
        when(talkStep.calculateMsgLength(user1)).thenReturn(10);` and write a similar one for countMatchKeywords()
The purpose of this is to test ButtonStep independently of TalkStep -- even if TalkStep's method fails, it won't be "blamed" on ButtonStep. Hope that makes sense! Ekagra went over it during our call today,  you can visit[ this link](https://semaphoreci.com/community/tutorials/stubbing-and-mocking-with-mockito-2-and-junit) he shared if u want

Lemiex reviewed (changes requested):
Sorry I realised there were a few things I forgot to cover, these all address ButtonStepTests:
1. the trace_type should be either "request" - user, "response" - chat
2. in init(), I think you got the userMsg and chatbotMsg for dialogue object mistaken: from the 3 user and 3 chtatbot messages, you should be able to make 3 dialogues:
dialogue1 has chat1 as chatbotMsg, user2 as userMsg
dialogue2 has chat 2 as chatbotMsg, user3 for userMsg
dialogue3 has chat3 as chatbotMsg, and empty userMsg
Recall that one dialogue is ONE back and forth, that starts when a chatbot speaks, and ends before the next chatbot message 
Also recall that we ignore the first user message in a conversation since there is no chatbot output to recommend. This is also documented in the parser

auroraborie committed:
Implemented TalkStep entities and tests to be compliant with ScoreStandard and fully tests TalkStep child entities.

auroraborie committed:
Changes transcript to separate dialogue objects.

auroraborie commented:
Hey Sarah 
thank you for all the feedback. I incorporated the changes. I also incorporated the 'max cap' feature we discussed in ConfidenceScoreOptimizer to address situations where the scoreAccumulator exceed the maxscore. You should find no further issue here.

Lemiex commented on backend/src/test/java/com/speakfluid/backend/entities/CardStepTests.java:
```diff
@@ -0,0 +1,176 @@
+    void init(){
+        card = new CardStep();
+        user1 = new WozMessage("SNG990", "i am looking for the most famous art gallery in this area." +
+                "can you tell me some suggestions?");
```
change trace_type

auroraborie replied:
oh good catch!

Lemiex reviewed (approved):
Beautiful, just one last change to implement I commented in the code :

auroraborie committed:
changed trace_type

auroraborie committed:
Merge branch 'main' into managertests2

auroraborie committed:
amended tests

auroraborie committed:
Merge branch 'managertests2' of https://github.com/edwardhuahan/tli-speakfluid into managertests2

auroraborie committed:
implemented ConfidenceScoreOptimizer and ConfidenceScoreCalculator Test Suites.

auroraborie committed:
saving changes

auroraborie committed:
implemented all tests for ButtonStep, CardStep, ConfidenceScoreOptimizer, ConfidenceScoreCalculator

auroraborie committed:
Merge branch 'main' into managertests2

auroraborie committed:
fix maven bugs

auroraborie committed:
Merge remote-tracking branch 'origin/managertests2' into managertests2

edwardhuahan commented:
Hey Aurora are you able to merge this branch with the main branch? I'm not sure why the Github Actions stopped working for this branch

edwardhuahan commented on backend/src/main/java/com/speakfluid/backend/entities/ButtonStep.java:
```diff
@@ -11,32 +11,36 @@
     public ButtonStep(){
         this.scoreAccumulator = 0.0;
     }
     private final List<Map<String, Double>> chatbotKeywordsScoreMap = Arrays.asList(
```
Is there a better way of accomplishing this rather than having all of these similar lines of code? It's not as easy to read the code. I think it would be great if there were more documentation

edwardhuahan replied:
Is there a better way of accomplishing this rather than having all of these similar lines of code? It's not as easy to read the code

edwardhuahan commented on backend/src/test/java/com/speakfluid/backend/entities/ButtonStepTests.java:
```diff
@@ -0,0 +1,171 @@
+    WozMessage user4;
+
+    // initialize array to an empty ArrayList otherwise default value is null.
+    ArrayList<WozMessage> userMsgs1 = new ArrayList<>();
```
Is there a reason that these have to remain separated? I think this could be better accomplished with an array instead of repeated code

edwardhuahan commented on backend/src/main/java/com/speakfluid/backend/usecases/WozWozTranscriptAnalysisInteractor.java:
```diff
@@ -70,19 +70,23 @@ public ArrayList<HashMap<String, ArrayList<Dialogue<WozMessage>>>> analyzeTransc
-                            dialogue.addStepSuggestion(stepSuggestion);
-                            dialogue.addConfidenceScore(confidenceScore);
+                        for(Map<String, Double> talkStepPair: suggestionPair) {
+                            for(Map.Entry<String, Double> entry : talkStepPair.entrySet()) {
```
Is it possible to implement this without adding another for loop? I'm aware that WozWozTranscriptAnalysisInteractor.java already contains multiple for loops

Lemiex commented on backend/src/main/java/com/speakfluid/backend/entities/ButtonStep.java:
```diff
@@ -45,20 +49,20 @@ public ButtonStep(){
-            if(chatbotMsgLength < 10){
-                scoreAccumulator += 2;
+            if(chatbotMsgLength < 10 && chatbotMsgLength > 2){
+                this.scoreAccumulator += 2;
```
use ScoreStandards.mediumWeight or .low

Lemiex commented on backend/src/main/java/com/speakfluid/backend/entities/ButtonStep.java:
```diff
@@ -45,20 +49,20 @@ public ButtonStep(){
             // if the user response is short, then this suggests buttons are suitable
-            if(userMsgLength <= 5){
+            if(userMsgLength <= 5 && userMsgLength > 1){
                 this.scoreAccumulator += 2;
```
use ScoreStandards.mediumWeight or .low, no magic numbers

Lemiex replied:
Same for all of your other step entities^^

Lemiex commented on backend/src/main/java/com/speakfluid/backend/entities/ConfidenceScoreCalculator.java:
```diff
@@ -21,20 +23,29 @@ public class ConfidenceScoreCalculator implements Scorable{
+        }
+        else{
+            double confidenceScore = (stepScoreAccumulator / stepTotalScore) * 100;
+            return Math.round(confidenceScore * 100.0) / 100.0;
```
why are we dividing by 100 here? doesn't Math.round(confidenceScore * 100.0) give us the confidence score in percentage already or am i doing the math wrong

Lemiex commented on backend/src/test/java/com/speakfluid/backend/entities/ButtonStepTests.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Test
+    public void testButtonRunAnalysisDialogue2(){
```
it might be helpful to name these methods as what they are testing, like testButtonRunAnalysisHighScore

Lemiex commented on backend/src/test/java/com/speakfluid/backend/entities/ConfidenceScoreOptimizerTests.java:
```diff
@@ -0,0 +1,288 @@
+
+        step1.put("Carousel", 68.75);
+        step2.put("Button", 66.67);
+        step3.put("Card", 20.0);
```
Nice! very thorough tests

Lemiex reviewed (approved):
I know you encountered problems with this passing the maven tests for this PR; glad it worked out in the end! very thorough testing altho more detailed test names would be helpful. Overall wonderful job
